File: StateMachine.hs
{-# LANGUAGE GADTs #-}
{-# LANGUAGE RankNTypes #-}

module StateMachine where

-- The same as Functor, but renamed to make it clear that we're
-- assuming it's polynomial
class Poly p where
  pmap :: (a -> b) -> p a -> p b

-- p × q
newtype CartesianProduct p q a = CartesianProduct (p a, q a)

instance (Poly p, Poly q) => Poly (CartesianProduct p q) where
  pmap f (CartesianProduct (p, q)) = CartesianProduct (pmap f p, pmap f q)

-- p ⊗ q
data DirichletProduct p q c where
  T :: p a -> q b -> (a -> b -> c) -> DirichletProduct p q c

instance (Poly p, Poly q) => Poly (DirichletProduct p q) where
  pmap f (T p q f') = T p q (\a b -> f (f' a b))

-- p ⊳ q
newtype SubstitutionProduct p q a = SubstitutionProduct (p (q a))

instance (Poly p, Poly q) => Poly (SubstitutionProduct p q) where
  pmap f (SubstitutionProduct p) = SubstitutionProduct (pmap (pmap f) p)

-- p^q
newtype CartesianClosure p q a = CartesianClosure (forall b. q b -> p (Either a b))

instance (Poly p) => Poly (CartesianClosure p q) where
  pmap f (CartesianClosure g) = CartesianClosure (\q -> pmap (either (Left . f) Right) (g q))

-- [p,q]
newtype DirichletClosure p q a = DirichletClosure (forall b. q b -> p (a, b))

instance (Poly p) => Poly (DirichletClosure p q) where
  pmap f (DirichletClosure g) = DirichletClosure (\q -> pmap (\(a, b) -> (f a, b)) (g q))

-- p × q^p → q
cartesianEval :: (Poly q) => CartesianProduct p (CartesianClosure q p) a -> q a
cartesianEval (CartesianProduct (p, CartesianClosure f)) = pmap (either id id) (f p)

-- p ⊗ [p,q] → q
dirichletEval :: (Poly q) => DirichletProduct p (DirichletClosure q p) a -> q a
dirichletEval (T p (DirichletClosure f) combine) = pmap (\(y, x) -> combine x y) (f p)

File: state_machine.ml
module type Poly = sig
  type 'a t

  val map : 'a t -> f:('a -> 'b) -> 'b t
end

(* p × q *)
module Cartesian_product (P : Poly) (Q : Poly) :
  Poly with type 'a t = 'a P.t * 'a Q.t = struct
  type 'a t = 'a P.t * 'a Q.t

  let map (p, q) ~f = (P.map p ~f, Q.map q ~f)
end

(* p ⊗ q *)
module Dirichlet_product (P : Poly) (Q : Poly) : sig
  type _ t = T : 'a P.t * 'b Q.t * ('a -> 'b -> 'c) -> 'c t

  include Poly with type 'a t := 'a t
end = struct
  type _ t = T : 'a P.t * 'b Q.t * ('a -> 'b -> 'c) -> 'c t

  let map (T (p, q, f')) ~f = T (p, q, fun a b -> f (f' a b))
end

(* p ⊳ q *)
module Substitution_product (P : Poly) (Q : Poly) :
  Poly with type 'a t = 'a Q.t P.t = struct
  type 'a t = 'a Q.t P.t

  let map p ~f = P.map p ~f:(fun q -> Q.map q ~f)
end

type ('a, 'b) either = Left of 'a | Right of 'b

(* p^q *)
module Cartesian_closure (P : Poly) (Q : Poly) : sig
  type 'a t = { f : 'b. 'b Q.t -> ('a, 'b) either P.t }

  include Poly with type 'a t := 'a t
end = struct
  type 'a t = { f : 'b. 'b Q.t -> ('a, 'b) either P.t }

  let map { f } ~f:g =
    {
      f =
        (fun q ->
          P.map (f q) ~f:(function Left a -> Left (g a) | Right b -> Right b));
    }
end

(* P × Q^P → Q *)
module Cartesian_eval (P : Poly) (Q : Poly) = struct
  module PQ = Cartesian_closure (Q) (P)
  module Prod = Cartesian_product (P) (PQ)

  let eval : type a. a Prod.t -> a Q.t =
   fun (p, { f }) -> Q.map (f p) ~f:(function Left a | Right a -> a)
end

(* [p,q] *)
module Dirichlet_closure (P : Poly) (Q : Poly) : sig
  type 'a t = { f : 'b. 'b Q.t -> ('a * 'b) P.t }

  include Poly with type 'a t := 'a t
end = struct
  type 'a t = { f : 'b. 'b Q.t -> ('a * 'b) P.t }

  let map { f } ~f:g =
    { f = (fun q -> P.map (f q) ~f:(fun (a, b) -> (g a, b))) }
end

(* P ⊗ [P,Q] → Q *)
module Dirichlet_eval (P : Poly) (Q : Poly) = struct
  module PQ = Dirichlet_closure (Q) (P)
  module Prod = Dirichlet_product (P) (PQ)

  let eval : type a. a Prod.t -> a Q.t =
   fun (T (p, { f }, combine)) -> Q.map (f p) ~f:(fun (y, x) -> combine x y)
end
